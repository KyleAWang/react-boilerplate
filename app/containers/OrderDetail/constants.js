export const LOAD_ORDER_DETAIL = 'boilerplate/OrderDetail/LOAD_ORDER_DETAIL';
export const LOAD_ORDER_DETAIL_SUCCESS = 'boilerplate/OrderDetail/LOAD_ORDER_DETAIL_SUCCESS';
export const LOAD_ORDER_DETAIL_ERROR = 'boilerplate/OrderDetail/LOAD_ORDER_DETAIL_ERROR';
