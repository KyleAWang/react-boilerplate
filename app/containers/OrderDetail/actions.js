import {LOAD_ORDER_DETAIL, LOAD_ORDER_DETAIL_ERROR, LOAD_ORDER_DETAIL_SUCCESS} from './constants';

export function loadOrderDetail() {
  return {
    type: LOAD_ORDER_DETAIL
  };
};

export function loadOrderDetailSuccess(order) {
  return {
    type: LOAD_ORDER_DETAIL_SUCCESS,
    order
  };
}

export function loadOrderDetailError(error) {
  return {
    type: LOAD_ORDER_DETAIL_ERROR,
    error
  };
};
