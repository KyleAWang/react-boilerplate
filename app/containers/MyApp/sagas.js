import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import request from 'utils/request';
import { LOAD_ORDERS } from './constants';
import { loadOrdersSuccess, loadOrdersError } from './actions';
// import { makeSelectOrders } from './selectors';


export function* getOrders() {
  const requestURL = 'http://localhost:3000/graphql?query={orders{orderId,totalCost,updated}}';

  try {
    const orders = yield call(request, requestURL);
    yield put(loadOrdersSuccess(orders.data.orders));
  } catch (err) {
    yield put(loadOrdersError(err));
  }
}

export function* ordersData() {
  const watcher = yield takeLatest(LOAD_ORDERS, getOrders);

  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

export default [
  ordersData,
];

