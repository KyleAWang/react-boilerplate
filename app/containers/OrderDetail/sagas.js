import { take, put, call, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import request from 'utils/request';
import { LOAD_ORDER_DETAIL } from './constants';
import { loadOrderDetailSuccess, loadOrderDetailError } from './actions';


export function* getOrder(action) {
  const requestUrl = `http://localhost:3000/graphql?query={order(orderId:"${action.orderId}"){orderId,totalCost,updated}}`;

  try {
    const order = yield call(request, requestUrl);
    yield put(loadOrderDetailSuccess(order.data.order));
  } catch (err) {
    yield put(loadOrderDetailError(err));
  }
}

export function* orderData() {
  const watcher = yield takeLatest(LOAD_ORDER_DETAIL, getOrder);

  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

export default [
  orderData,
];
