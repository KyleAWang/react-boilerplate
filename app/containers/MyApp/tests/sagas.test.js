import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';
import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_ORDERS } from '../constants';
import { loadOrdersSuccess, loadOrdersError } from '../actions';
import { getOrders, ordersData } from '../sagas';

describe('getOrders Saga', () => {
  let getOrdersGenerator;

  beforeEach(() => {
    getOrdersGenerator = getOrders();

    const callDescriptor = getOrdersGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadOrdersSuccess action if it requests the data successfully', () => {
    const response = {
      data: {
        orders: [{
          orderId: '333',
        }, {
          orderId: '444',
        }],
      },
    };
    const putDescriptor = getOrdersGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadOrdersSuccess(response.data.orders)));
  });

  it('should call the loadOrdersError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getOrdersGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadOrdersError(response)));
  });
});

describe('ordersData Saga', () => {
  const ordersDataSaga = ordersData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_ORDERS action', () => {
    const takeLatestDescriptor = ordersDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_ORDERS, getOrders));
  });

  it('should yield until loaction change action', () => {
    const takeDescriptior = ordersDataSaga.next(mockedTask).value;
    expect(takeDescriptior).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = ordersDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
