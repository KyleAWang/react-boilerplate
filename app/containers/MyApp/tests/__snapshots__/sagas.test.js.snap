exports[`getOrders Saga should call the loadOrdersError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3000/graphql?query={orders{orderId,totalCost,updated}}",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getOrders Saga should dispatch the loadOrdersSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3000/graphql?query={orders{orderId,totalCost,updated}}",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
